src/shared/port-provider.js

```diff
@@ -0,0 +1,269 @@
+ *
+ *
+ *                 PortProvider                                       |                PortFinder
+ * -------------------------------------------------------------------|-------------------------------------------------------------
+ *
+ * 2. listens to requests of `MessagePort` <---------------------------- 1. request a `MessagePort` using `window.parent.postMessage`
+ *                 |
+ *                 V
+ * 3. sends offers of `MessagePort` using `event.source.postMessage` ---> 4. listens to offers of `MessagePort`
+ *                 |
+ *                 V
+ * 5. send reciprocal port to the `sidebar` frame using the
+ *    `host-sidebar(MessageChannel).sidebar(MessagePort)`
+ *
+ *
+ * In some situations, because `guest` iframe/s load in parallel to the `host`
+ * frame, we can not assume that the code in the `host` frame is executed before
+ * the code in a `guest` frame. Hence, we can't assume that `PortProvider` (in
+ * the `host` frame) is initialized before `PortFinder` (in the `guest` frame).
+ * Therefore, for the `PortFinder`, we implement a polling strategy (sending a
+ * message every N milliseconds) until a response is received.
+ *
+ * Channel nomenclature is `[frame1]-[frame2]` so that:
+ *   - `port1` should be owned by/transferred to `frame1`, and
+ *   - `port2` should be owned by/transferred to `frame2`
+ *
+ * @implements Destroyable
+ */
+export class PortProvider {
+  /**
+   * @param {string} hypothesisAppsURL
+   */
+  constructor(hypothesisAppsURL) {
+    this._hypothesisAppsOrigin = new URL(hypothesisAppsURL).origin;
+
+    // Although some channels (v.gr. `notebook-sidebar`) have only one
+    // `MessageChannel`, other channels (v.gr. `guest-sidebar`) can have multiple
+    // `MessageChannel`s. In spite of the number channel, we store all
+    // `MessageChannel` on a `Map<Window, MessageChannel>`.
+    /** @type {Map<Channel, Map<Window, MessageChannel>>} */
+    this._channels = new Map();
+
+    // Create the `host-sidebar` channel immediately, while other channels are
+    // created on demand
+    this._hostSidebarChannel = new MessageChannel();
+    this._channels.set(
+      'host-sidebar',
+      new Map([[window, this._hostSidebarChannel]])
+    );
+
+    this._listeners = new ListenerCollection();
+    this._listen();
+  }
+
+  /**
+   * @param {'onHostPortRequest'} _eventName
+   * @param {(MessagePort, channel: 'guest') => void} handler - this handler
+   *   fires when a request for the 'guest-host' channel is listened.
+   */
+  addEventListener(_eventName, handler) {
+    this._onHostPortRequest = handler;
+  }
+
+  /**
+   * Returns a port from a channel. Currently, only returns the `host` port from
+   * the `host-Sidebar` channel. Otherwise, it returns `null`.
+   *
+   * @param {object} options
+   *   @param {'host-sidebar'} options.channel
+   *   @param {'host'} options.port
+   */
+  getPort({ channel, port }) {
+    if (channel === 'host-sidebar' && port === 'host') {
```

Comment:
Presumably a hypothetical `host-notebook` channel is another possible future use case for this method.

Reply:
Yes, you are correct.